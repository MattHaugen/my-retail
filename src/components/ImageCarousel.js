import React from 'react';
import styled from 'styled-components';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import faSearchPlus from '@fortawesome/fontawesome-free-solid/faSearchPlus';
import faChevronLeft from '@fortawesome/fontawesome-free-solid/faChevronLeft';
import faChevronRight from '@fortawesome/fontawesome-free-solid/faChevronRight';
import colors from '../constants/colors';

const CarouselContainer = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
  flex-direction: row;
  flex-wrap: wrap;
`;

const ZoomControl = styled.button`
  width: 100%;
  color: ${colors.darkGray};
  text-align: center;
  font-size: 14px;
  margin-top: 15px;
  border: none;
  background-color: #FFF;
  cursor: pointer;

  .image-carousel-zoom-icon {
    margin-right: 5px;
  }
`;

const CarouselNavigation = styled.div`
  width: 100%;
  height: 70px;

  button {
    height: 100%;

    img {
      height: 100%;
    }
  }
`;

class ImageCarousel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      images: [...props.images.PrimaryImage, ...props.images.AlternateImages],
      primaryImageIndex: 0
    }
  }

  getNextImageIndex(currentIndex, numberOfImages) {
    return currentIndex + 1 === numberOfImages ? 0 : currentIndex + 1;
  }

  getPreviousImageIndex(currentIndex, numberOfImages) {
    return currentIndex - 1 === -1 ? numberOfImages - 1 : currentIndex - 1;
  }

  selectImage(imageIndex) {
    this.setState({
      primaryImageIndex: imageIndex
    });
  }

  render() {

    let classReference = this;
    let nextImageIndex = this.getNextImageIndex(this.state.primaryImageIndex, this.state.images.length);
    let previousImageIndex = this.getPreviousImageIndex(this.state.primaryImageIndex, this.state.images.length)
    let previewImageIndexes = [ previousImageIndex, this.state.primaryImageIndex, nextImageIndex ];

    const previewImages = previewImageIndexes.map((imageIndex) => {
      return (
        <button key={imageIndex} onClick={() => classReference.selectImage(imageIndex)}>
          <img src={this.state.images[imageIndex].image} alt='Preview' />
        </button>
      )
    });

    return (
      <CarouselContainer>
        <img src={this.state.images[this.state.primaryImageIndex].image} alt='Primary'/>
        <ZoomControl>
          <FontAwesomeIcon icon={faSearchPlus} className='image-carousel-zoom-icon'/>view larger
        </ZoomControl>
        <CarouselNavigation>
          <button onClick={() => this.selectImage(previousImageIndex)}><FontAwesomeIcon icon={faChevronLeft} /></button>
          {previewImages}
          <button onClick={() => this.selectImage(nextImageIndex)}><FontAwesomeIcon icon={faChevronRight} /></button>
        </CarouselNavigation>
      </CarouselContainer>
    )
  }
}

export default ImageCarousel;
