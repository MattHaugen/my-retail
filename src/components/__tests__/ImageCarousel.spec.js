import React from 'react';
import ImageCarousel from '../ImageCarousel';
import renderer from 'react-test-renderer';

describe('ImageCarousel component', () => {
  it('properly renders initial state', () => {
    //ARRANGE
    const inputImages = {
      AlternateImages: [
        {
          image: "Alternate 1"
        },
        {
          image: "Alternate 2"
        },
        {
          image: "Alternate 3"
        }
      ],
      PrimaryImage: [
        {
          image: "Primary Image"
        }
      ]
    };

    //ACT
    const component = renderer.create(<ImageCarousel images={inputImages} />);
    const json = component.toJSON();

    //ASSERT
    expect(json).toMatchSnapshot();
  });

  describe('selectImage', () => {
    it('properly sets primaryImageIndex', () => {
      //ARRANGE
      const inputImages = {
        AlternateImages: [
          {
            image: "Alternate 1"
          },
          {
            image: "Alternate 2"
          },
          {
            image: "Alternate 3"
          }
        ],
        PrimaryImage: [
          {
            image: "Primary Image"
          }
        ]
      };

      //ACT
      const component = renderer.create(<ImageCarousel images={inputImages} />);
      component.getInstance().selectImage(2);

      //ASSERT
      expect(component.getInstance().state.primaryImageIndex).toEqual(2);
    });
  });

  describe('getNextImageIndex', () => {
    it('returns zero if we are at the end of the array of images', () => {
      //ARRANGE
      const inputImages = {
        AlternateImages: [
          {
            image: "Alternate 1"
          },
          {
            image: "Alternate 2"
          }
        ],
        PrimaryImage: [
          {
            image: "Primary Image"
          }
        ]
      };

      //ACT
      const component = renderer.create(<ImageCarousel images={inputImages} />);
      let result = component.getInstance().getNextImageIndex(2, 3);

      //ASSERT
      expect(result).toEqual(0);
    });

    it('returns the next index otherwise', () => {
      //ARRANGE
      const inputImages = {
        AlternateImages: [
          {
            image: "Alternate 1"
          },
          {
            image: "Alternate 2"
          }
        ],
        PrimaryImage: [
          {
            image: "Primary Image"
          }
        ]
      };

      //ACT
      const component = renderer.create(<ImageCarousel images={inputImages} />);
      let result = component.getInstance().getNextImageIndex(1, 3);

      //ASSERT
      expect(result).toEqual(2);
    });
  });

  describe('getPreviousImageIndex', () => {
    it('returns the last index if we are at the beginning of the array of images', () => {
      //ARRANGE
      const inputImages = {
        AlternateImages: [
          {
            image: "Alternate 1"
          },
          {
            image: "Alternate 2"
          }
        ],
        PrimaryImage: [
          {
            image: "Primary Image"
          }
        ]
      };

      //ACT
      const component = renderer.create(<ImageCarousel images={inputImages} />);
      let result = component.getInstance().getPreviousImageIndex(0, 3);

      //ASSERT
      expect(result).toEqual(2);
    });

    it('returns the previous index otherwise', () => {
      //ARRANGE
      const inputImages = {
        AlternateImages: [
          {
            image: "Alternate 1"
          },
          {
            image: "Alternate 2"
          }
        ],
        PrimaryImage: [
          {
            image: "Primary Image"
          }
        ]
      };

      //ACT
      const component = renderer.create(<ImageCarousel images={inputImages} />);
      let result = component.getInstance().getPreviousImageIndex(0, 3);

      //ASSERT
      expect(result).toEqual(2);
    });
  });

});
