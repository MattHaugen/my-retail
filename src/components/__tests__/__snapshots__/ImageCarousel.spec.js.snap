// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCarousel component properly renders initial state 1`] = `
<div
  className="sc-bdVaJa drgVKh"
>
  <img
    alt="Primary"
    className="image-carousel-primary-image"
    src="Primary Image"
  />
  <button
    className="sc-bwzfXH kiSqPT"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-search-plus fa-w-16 image-carousel-zoom-icon"
      data-icon="search-plus"
      data-prefix="fas"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 192v32c0 6.6-5.4 12-12 12h-56v56c0 6.6-5.4 12-12 12h-32c-6.6 0-12-5.4-12-12v-56h-56c-6.6 0-12-5.4-12-12v-32c0-6.6 5.4-12 12-12h56v-56c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v56h56c6.6 0 12 5.4 12 12zm201 284.7L476.7 505c-9.4 9.4-24.6 9.4-33.9 0L343 405.3c-4.5-4.5-7-10.6-7-17V372c-35.3 27.6-79.7 44-128 44C93.1 416 0 322.9 0 208S93.1 0 208 0s208 93.1 208 208c0 48.3-16.4 92.7-44 128h16.3c6.4 0 12.5 2.5 17 7l99.7 99.7c9.3 9.4 9.3 24.6 0 34zM344 208c0-75.2-60.8-136-136-136S72 132.8 72 208s60.8 136 136 136 136-60.8 136-136z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    view larger
  </button>
  <div
    className="sc-htpNat fKtlMk"
  >
    <button
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-chevron-left fa-w-10 "
        data-icon="chevron-left"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
    <button
      onClick={[Function]}
    >
      <img
        alt="Preview"
        src="Alternate 3"
      />
    </button>
    <button
      onClick={[Function]}
    >
      <img
        alt="Preview"
        src="Primary Image"
      />
    </button>
    <button
      onClick={[Function]}
    >
      <img
        alt="Preview"
        src="Alternate 1"
      />
    </button>
    <button
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-chevron-right fa-w-10 "
        data-icon="chevron-right"
        data-prefix="fas"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </div>
</div>
`;
